### PR TITLE
feat(recipe): add Super 3.5 VL continued pretraining on FineWeb

### PR DESCRIPTION
**Pull request by HuiyingLi**
# What does this PR do?

Add a reproducible FineWeb continued-pretraining recipe and tutorial for Nemotron 3.5 Super VL, with a W&B screenshot of 200 raw training-loss measurements and online tracking.

# Changelog

- Add an 8-node, 64-H100 recipe for full-parameter training of the language backbone using FSDP2, CP4, EP64, 4,096-token sequences, and a global batch of 32. The vision tower and projector remain frozen.
- Document every data-preparation step: pin the FineWeb and model/tokenizer revisions, download the Parquet shard, select 100,000 documents, save tokenizer assets, run Megatron preprocessing with appended EOD tokens, and verify document/token counts, tokenization round-trip, and `.bin/.idx` checksums.
- Explain the document-level train/validation split, Slurm launch, online W&B configuration, checkpoint/resume usage, and the W&B training-loss chart. Include the 200-step training-loss screenshot with smoothing disabled and eight held-out validation measurements in the result table.
- Preserve Mamba FP32 parameter storage during Omni initialization so checkpoint restoration can use native TE FusedAdam state initialization. Add focused regression coverage.

# Before this PR is ready for review

- [x] Read and followed the contributor guidelines.
- [x] Added necessary regression tests.
- [x] Added the recipe, tutorial, data-generation commands, and result artifacts.
- [x] Added DCO sign-off to every commit.

# Validation

- DFW, 64 H100 GPUs: 200 completed optimizer updates and eight validation passes; all training and validation losses checked against unsampled W&B history.
- FineWeb preparation: 100,000 documents and 69,756,226 tokens; indexed-file checksums and tokenizer round-trip verified by running the tutorial's inspection code.
- DFW, 8 H100 GPUs: six updates on a four-layer pretrained slice with CP4, EP8, and TE FusedAdam; restore after update three and complete three further updates using native optimizer state initialization. The resume probe uses `checkpoint.cpu_offload=false`. Model, optimizer, and scheduler state match exactly on every rank at restore; subsequent training loss differs from the previous wrapper control by at most 4.1e-5. Both runs finished and their online W&B histories match the training logs.
- DFW, 8 H100 GPUs: the existing main training startup completed two FineWeb updates on a four-layer pretrained slice (CP4/EP8) launched as direct Slurm ranks; finite losses and gradients, clean exit, and finished online W&B history verified against the training log.
- CPU Omni and optimizer-state tests: 23 passed. The new sharded FP32 regression fails before the initialization fix and passes afterward.
- Ruff formatting/lint checks passed. `make docs-check` validated 482 MDX pages with 0 errors (2 warnings).

### CODE REVIEW

**Review by akoumpa — DISMISSED** [review]

File: examples/llm_pretrain/nemotron_3_5_super_vl/nemotron_3_5_super_vl_fineweb_cpt.yaml [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  cpu_offload: true
+
+loss_fn:
+  _target_: nemo_automodel.components.loss.masked_ce.MaskedCrossEntropy
+  fp32_upcast: true
+
+dataset:
+  _target_: nemo_automodel.components.datasets.llm.megatron_dataset.MegatronPretraining
+  paths: ${oc.env:FINEWEB_DATA_DIR}/fineweb_0_text_document
+  index_mapping_dir: ${oc.env:FINEWEB_DATA_DIR}/index_mapping
```
Comment (akoumpa) [inline examples/llm_pretrain/nemotron_3_5_super_vl/nemotron_3_5_super_vl_fineweb_cpt.yaml]: @HuiyingLi we don't support oc.env IIRC

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 2]
```diff
@@ -0,0 +1,490 @@
+---
+title: "Nemotron 3.5 Super VL: Continued Pretraining on FineWeb"
+description: "Prepare FineWeb as Megatron indexed data, continue training the Super 3.5 VL language backbone, and track training and validation loss with Weights & Biases."
+position: 17
+---
+
+Continue training the language backbone of **Nemotron 3.5 Super VL** on raw
+[FineWeb](https://huggingface.co/datasets/HuggingFaceFW/fineweb) text. This guide follows the
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
This guide shows how to continue training the language backbone of **Nemotron 3.5 Super VL** on raw
[FineWeb](https://huggingface.co/datasets/HuggingFaceFW/fineweb) text. It follows the
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+The model is loaded from
+`nvidia/NVIDIA-Nemotron-3.5-Super-midtrain-67B-vision-pretrained`. Despite the historical
+`67B` in its name, this checkpoint contains approximately 121B parameters. The complete
+VL model is retained; its RADIO vision tower and vision projector are frozen and unused
+during this text-only run. The language backbone is trained with full-parameter updates,
+with no LoRA adapters. The MTP head is disabled.
+
+## What This Run Measures
+
+The recipe uses the first 100,000 documents of one shard in FineWeb's `sample/10BT`
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
The recipe uses the first 100,000 documents from one shard in FineWeb's `sample/10BT`
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+labels:    [t1, t2, ..., t4096]
+loss_mask: [ 1,  1, ...,     1]
+```
+
+The model learns to predict all next tokens, including document-ending EOS tokens.
+Validation loss is computed on held-out FineWeb text using the same tokenizer and objective.
+A lower loss on this sample does not establish improved downstream task accuracy or retained
+image understanding.
+
+## Hardware and Run Configuration
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
## Guide Overview

| Step | Description |
|---|---|
| **Step 0** | Set up the environment |
| **Step 1** | Download a reproducible FineWeb sample |
| **Step 2** | Tokenize the sample into Megatron `.bin/.idx` files |
| **Step 3** | Review the continued-pretraining YAML |
| **Step 4** | Launch training with Slurm and online W&B |
| **Step 5** | Inspect the checkpoint and resume training |
| **Step 6** | View the training loss in W&B |

## Hardware Requirements

- **8 nodes × 8 H100 80 GB GPUs** (64 GPUs) required for this recipe
- **Storage**: Approximately 232 GiB for the base BF16 checkpoint, and approximately 6 TiB of additional shared storage for checkpoint retention

## Run Configuration
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+| Setting | Value |
+|---|---|
+| GPUs | 8 nodes × 8 H100 80 GB |
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+print(json.dumps(manifest["megatron"], indent=2))
+```
+
+Run the script from the same shell where you exported the shared paths:
+
+```bash
+python inspect_fineweb.py
+```
+
+Expected counts are `100000` documents and `69756226` tokens. The manifest now
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
Expected counts are `100,000` documents and `69,756,226` tokens. The manifest now
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+checksums of the indexed training files. Set `FINEWEB_DATA_DIR` to this same shared
+directory on every training node.
+
+## Step 3 — Review the Continued-Pretraining YAML
+
+Use
+[`nemotron_3_5_super_vl_fineweb_cpt.yaml`](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_pretrain/nemotron_3_5_super_vl/nemotron_3_5_super_vl_fineweb_cpt.yaml).
+
+The `TrainFinetuneRecipeForNextTokenPrediction` recipe is also the existing LLM pretraining
+entrypoint. Its name does not imply an SFT objective: `MegatronPretraining` supplies raw,
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
entry point. Its name does not imply an SFT objective: `MegatronPretraining` supplies raw,
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+shifted next-token targets. `from_pretrained` loads the existing weights rather than
+randomly initializing a model. Architecture registration resolves the causal-LM entrypoint
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
randomly initializing a model. Architecture registration resolves the causal-LM entry point
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+Use the original sharded training checkpoint with `checkpoint.restore_from` to resume optimizer,
+scheduler, and dataloader state. Add the following override to the per-rank training
+command after checking that `LATEST` resolves to a complete checkpoint:
+
+```bash
+--checkpoint.restore_from=/path/to/checkpoints/LATEST
+```
+
+Keep the tokenizer, dataset prefix and revision, train/validation split, seed, and batch
+semantics fixed when resuming training. The optimizer scheduler restores its
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
Keep the tokenizer, dataset prefix, dataset revision, training and validation split, seed, and batch
semantics fixed when resuming training. The learning-rate scheduler restores its
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+schedule when continuing training.
+
+## Step 6 — View the Training Loss in W&B
+
+Open your W&B run and select **Charts**. In the `loss` panel, use **Step**
+for the X axis and `loss` for the Y axis. Set **Smoothing** to `0` in the panel's
+settings. Open the panel in full-screen view to inspect all 200 training measurements
+and capture the chart for the tutorial.
+
+## Measured Results — 200 Steps
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
## Measured Results for 200 Steps
```

File: docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+title: "Nemotron 3.5 Super VL: Continued Pretraining on FineWeb"
+description: "Prepare FineWeb as Megatron indexed data, continue training the Super 3.5 VL language backbone, and track training and validation loss with Weights & Biases."
+position: 17
+---
+
+Continue training the language backbone of **Nemotron 3.5 Super VL** on raw
+[FineWeb](https://huggingface.co/datasets/HuggingFaceFW/fineweb) text. This guide follows the
+setup, training, checkpoint, and results structure of the
+[CORD-v2 tutorial](/recipes-e2e-examples/nemotron-3-5-super-vl), using a next-token pretraining
+objective and **Megatron indexed `.bin/.idx` data**.
```
Comment (jgerh) [inline docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx]:
```suggestion
[CORD-v2 tutorial](/recipes-e2e-examples/nemotron-3-5-super-vl) and uses a next-token pretraining
objective and **Megatron-indexed `.bin/.idx` data**.
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of docs/guides/vlm/nemotron-3-5-super-vl-continued-pretraining.mdx. Reorganized the overview content so it is similar to the VLM guide pattern used in some files. Added a Guide Overview section, separated hardware requirements into short bullets, and moved recipe-specific settings into a dedicated Run Configuration table. Also provided some copyedits. Please review changes and let us know if this new reorg works for you.

**Review by akoumpa — APPROVED** [review]
